File: PAT/coding/67.matrix.py
import sys
from functools import cache


sys.setrecursionlimit(5000)

n = int(input())
m = list(map(int, input().split()))

matrix = []
for i in range(n):
    matrix.append([m[i], m[i + 1]])

cut = [[0] * (n + 1) for _ in range(n + 1)]


@cache
def dfs(i, j):
    """Return the minimum multiplication cost for matrices i through j."""
    if i == j:
        return 0

    res = 999999
    for k in range(i, j):
        cost = dfs(i, k) + dfs(k + 1, j) + m[i - 1] * m[k] * m[j]
        if cost < res:
            res = cost
            cut[i][j] = k

    return res


@cache
def build(i, j):
    if i == j:
        return f"M{i}"

    k = cut[i][j]
    return "(" + build(i, k) + ")x(" + build(k + 1, j) + ")"


res = dfs(1, n)
print(build(1, n))
print(res)

File: PAT/coding/67.matrix.cpp
#include <bits/stdc++.h>
using namespace std;

int n;
vector<long long> m;
vector<vector<long long>> dp;
vector<vector<int>> cut;

long long dfs(int i, int j) {
    if (i == j) {
        return 0;
    }

    // 已经计算过
    if (dp[i][j] != -1) {
        return dp[i][j];
    }

    long long res = LLONG_MAX;

    for (int k = i; k < j; k++) {
        long long cost =
            dfs(i, k)
            + dfs(k + 1, j)
            + m[i - 1] * m[k] * m[j];

        if (cost < res) {
            res = cost;
            cut[i][j] = k;
        }
    }

    return dp[i][j] = res;
}

string build(int i, int j) {
    if (i == j) {
        return "M" + to_string(i);
    }

    int k = cut[i][j];

    return "(" + build(i, k) + ")x(" + build(k + 1, j) + ")";
}

int main() {
    cin >> n;

    m.resize(n + 1);
    for (int i = 0; i <= n; i++) {
        cin >> m[i];
    }

    dp.assign(n + 1, vector<long long>(n + 1, -1));
    cut.assign(n + 1, vector<int>(n + 1, 0));

    long long res = dfs(1, n);

    cout << build(1, n) << endl;
    cout << res << endl;

    return 0;
}